File: japiwzsk/M2msessWzsk.java
/**
  * \file M2msessWzsk.java
  * Java API code for job M2msessWzsk
	* \copyright (C) 2018-2020 MPSI Technologies GmbH
	* \author Emily Johnson (auto-generation)
	* \date created: 5 Dec 2020
	*/
// IP header --- ABOVE

package apiwzsk;

import java.util.*;
import org.w3c.dom.*;
import sbecore.*;

public class M2msessWzsk {
	/**
	  * StatShr (full: StatShrM2msessWzsk)
	  */
	public class StatShr extends Block {

		public static final int SCRJREFACQPREVIEW = 1;
		public static final int SCRJREFACQPTCLOUD = 2;
		public static final int SCRJREFACTEXPOSURE = 3;
		public static final int SCRJREFACTLASER = 4;
		public static final int SCRJREFACTSERVO = 5;
		public static final int SCRJREFIPRCORNER = 6;
		public static final int SCRJREFIPRTRACE = 7;
		public static final int SCRJREFSRCSYSINFO = 8;

		public StatShr(
					String scrJrefAcqpreview
					, String scrJrefAcqptcloud
					, String scrJrefActexposure
					, String scrJrefActlaser
					, String scrJrefActservo
					, String scrJrefIprcorner
					, String scrJrefIprtrace
					, String scrJrefSrcsysinfo
				) {
			this.scrJrefAcqpreview = scrJrefAcqpreview;
			this.scrJrefAcqptcloud = scrJrefAcqptcloud;
			this.scrJrefActexposure = scrJrefActexposure;
			this.scrJrefActlaser = scrJrefActlaser;
			this.scrJrefActservo = scrJrefActservo;
			this.scrJrefIprcorner = scrJrefIprcorner;
			this.scrJrefIprtrace = scrJrefIprtrace;
			this.scrJrefSrcsysinfo = scrJrefSrcsysinfo;

			mask = new HashSet<Integer>(Arrays.asList(SCRJREFACQPREVIEW, SCRJREFACQPTCLOUD, SCRJREFACTEXPOSURE, SCRJREFACTLASER, SCRJREFACTSERVO, SCRJREFIPRCORNER, SCRJREFIPRTRACE, SCRJREFSRCSYSINFO));
		};

		public String scrJrefAcqpreview;
		public String scrJrefAcqptcloud;
		public String scrJrefActexposure;
		public String scrJrefActlaser;
		public String scrJrefActservo;
		public String scrJrefIprcorner;
		public String scrJrefIprtrace;
		public String scrJrefSrcsysinfo;

		public boolean readXML(
					Document doc
					, String basexpath
					, boolean addbasetag
				) {

			clear();

			if (addbasetag) basexpath = Xmlio.checkUclcXPaths(doc, basexpath, "StatShrM2msessWzsk");

			String itemtag = "StatitemShrM2msessWzsk";

			if (Xmlio.checkXPath(doc, basexpath)) {
				scrJrefAcqpreview = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Si", "sref", "scrJrefAcqpreview", mask, SCRJREFACQPREVIEW);
				scrJrefAcqptcloud = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Si", "sref", "scrJrefAcqptcloud", mask, SCRJREFACQPTCLOUD);
				scrJrefActexposure = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Si", "sref", "scrJrefActexposure", mask, SCRJREFACTEXPOSURE);
				scrJrefActlaser = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Si", "sref", "scrJrefActlaser", mask, SCRJREFACTLASER);
				scrJrefActservo = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Si", "sref", "scrJrefActservo", mask, SCRJREFACTSERVO);
				scrJrefIprcorner = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Si", "sref", "scrJrefIprcorner", mask, SCRJREFIPRCORNER);
				scrJrefIprtrace = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Si", "sref", "scrJrefIprtrace", mask, SCRJREFIPRTRACE);
				scrJrefSrcsysinfo = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Si", "sref", "scrJrefSrcsysinfo", mask, SCRJREFSRCSYSINFO);

				return true;
			};

			return false;
		};

		public HashSet<Integer> comm(
					StatShr comp
				) {
			HashSet<Integer> items = new HashSet<Integer>();

			if (scrJrefAcqpreview.equals(comp.scrJrefAcqpreview)) items.add(SCRJREFACQPREVIEW);
			if (scrJrefAcqptcloud.equals(comp.scrJrefAcqptcloud)) items.add(SCRJREFACQPTCLOUD);
			if (scrJrefActexposure.equals(comp.scrJrefActexposure)) items.add(SCRJREFACTEXPOSURE);
			if (scrJrefActlaser.equals(comp.scrJrefActlaser)) items.add(SCRJREFACTLASER);
			if (scrJrefActservo.equals(comp.scrJrefActservo)) items.add(SCRJREFACTSERVO);
			if (scrJrefIprcorner.equals(comp.scrJrefIprcorner)) items.add(SCRJREFIPRCORNER);
			if (scrJrefIprtrace.equals(comp.scrJrefIprtrace)) items.add(SCRJREFIPRTRACE);
			if (scrJrefSrcsysinfo.equals(comp.scrJrefSrcsysinfo)) items.add(SCRJREFSRCSYSINFO);

			return(items);
		};

		public HashSet<Integer> diff(
					StatShr comp
				) {
			HashSet<Integer> commitems;
			HashSet<Integer> diffitems;

			commitems = comm(comp);

			diffitems = new HashSet<Integer>(Arrays.asList(SCRJREFACQPREVIEW, SCRJREFACQPTCLOUD, SCRJREFACTEXPOSURE, SCRJREFACTLASER, SCRJREFACTSERVO, SCRJREFIPRCORNER, SCRJREFIPRTRACE, SCRJREFSRCSYSINFO));
			for (Integer ci: commitems) diffitems.remove(ci);

			return(diffitems);
		};

	};

	/**
		* DpchEngData (full: DpchEngM2msessWzskData)
		*/
	public class DpchEngData extends DpchEngWzsk {

		public static final int SCRJREF = 1;
		public static final int STATSHR = 2;

		public DpchEngData() {
			super(VecWzskVDpch.DPCHENGM2MSESSWZSKDATA);

			statshr = new StatShr("", "", "", "", "", "", "", "");
		};

		public StatShr statshr;

		public String getSrefsMask() {
			ArrayList<String> ss = new ArrayList<String>();

			if (has(SCRJREF)) ss.add("scrJref");
			if (has(STATSHR)) ss.add("statshr");

			return StrMod.vectorToString(ss, ';');
		};

		public void readXML(
					Document doc
					, String basexpath
					, boolean addbasetag
				) {

			clear();

			if (addbasetag) basexpath = Xmlio.checkUclcXPaths(doc, basexpath, "DpchEngM2msessWzskData");

			if (Xmlio.checkXPath(doc, basexpath)) {
				scrJref = Xmlio.extractStringUclc(doc, basexpath, "scrJref", "", mask, SCRJREF);
				if (statshr.readXML(doc, basexpath, true)) add(STATSHR);
			} else {
				scrJref = "";
				statshr = new StatShr("", "", "", "", "", "", "", "");
			};
		};

	};

};

File: sapiwzsk/M2msessWzsk.swift
/**
  * \file M2msessWzsk.swift
  * Swift API code for job M2msessWzsk
	* \copyright (C) 2022 MPSI Technologies GmbH
	* \author Alexander Wirthmueller (auto-generation)
	* \date created: 25 Mar 2022
	*/
// IP header --- ABOVE

import Foundation
import sbecore

/**
	* M2msessWzsk
	*/
public class M2msessWzsk {
	/**
	  * StatShr (full: StatShrM2msessWzsk)
	  */
	public class StatShr: Block {

		public static let SCRJREFACQPREVIEW: Int = 1
		public static let SCRJREFACQPTCLOUD: Int = 2
		public static let SCRJREFACTEXPOSURE: Int = 3
		public static let SCRJREFACTLASER: Int = 4
		public static let SCRJREFACTSERVO: Int = 5
		public static let SCRJREFIPRCORNER: Int = 6
		public static let SCRJREFIPRTRACE: Int = 7
		public static let SCRJREFSRCSYSINFO: Int = 8

		public init(
					_ scrJrefAcqpreview: String = ""
					, _ scrJrefAcqptcloud: String = ""
					, _ scrJrefActexposure: String = ""
					, _ scrJrefActlaser: String = ""
					, _ scrJrefActservo: String = ""
					, _ scrJrefIprcorner: String = ""
					, _ scrJrefIprtrace: String = ""
					, _ scrJrefSrcsysinfo: String = ""
				) {
			self.scrJrefAcqpreview = scrJrefAcqpreview
			self.scrJrefAcqptcloud = scrJrefAcqptcloud
			self.scrJrefActexposure = scrJrefActexposure
			self.scrJrefActlaser = scrJrefActlaser
			self.scrJrefActservo = scrJrefActservo
			self.scrJrefIprcorner = scrJrefIprcorner
			self.scrJrefIprtrace = scrJrefIprtrace
			self.scrJrefSrcsysinfo = scrJrefSrcsysinfo

			super.init()

			mask = [StatShr.SCRJREFACQPREVIEW, StatShr.SCRJREFACQPTCLOUD, StatShr.SCRJREFACTEXPOSURE, StatShr.SCRJREFACTLASER, StatShr.SCRJREFACTSERVO, StatShr.SCRJREFIPRCORNER, StatShr.SCRJREFIPRTRACE, StatShr.SCRJREFSRCSYSINFO]
		}

		public var scrJrefAcqpreview: String
		public var scrJrefAcqptcloud: String
		public var scrJrefActexposure: String
		public var scrJrefActlaser: String
		public var scrJrefActservo: String
		public var scrJrefIprcorner: String
		public var scrJrefIprtrace: String
		public var scrJrefSrcsysinfo: String

		public func readXML(
					_ doc: XMLDocument
					, _ _basexpath: String
					, _ addbasetag: Bool
				) -> Bool {
			var valid: Bool
			var basexpath: String? = _basexpath

			clear()

			if addbasetag {basexpath = Xmlio.checkUclcXPaths(doc, _basexpath, "StatShrM2msessWzsk")}
			valid = basexpath != nil

			let itemtag: String = "StatitemShrM2msessWzsk"

			if valid {
				scrJrefAcqpreview = Xmlio.extractStringAttrUclc(doc, basexpath!, itemtag, "Ci", "sref", "scrJrefAcqpreview", &mask, StatShr.SCRJREFACQPREVIEW)
				scrJrefAcqptcloud = Xmlio.extractStringAttrUclc(doc, basexpath!, itemtag, "Ci", "sref", "scrJrefAcqptcloud", &mask, StatShr.SCRJREFACQPTCLOUD)
				scrJrefActexposure = Xmlio.extractStringAttrUclc(doc, basexpath!, itemtag, "Ci", "sref", "scrJrefActexposure", &mask, StatShr.SCRJREFACTEXPOSURE)
				scrJrefActlaser = Xmlio.extractStringAttrUclc(doc, basexpath!, itemtag, "Ci", "sref", "scrJrefActlaser", &mask, StatShr.SCRJREFACTLASER)
				scrJrefActservo = Xmlio.extractStringAttrUclc(doc, basexpath!, itemtag, "Ci", "sref", "scrJrefActservo", &mask, StatShr.SCRJREFACTSERVO)
				scrJrefIprcorner = Xmlio.extractStringAttrUclc(doc, basexpath!, itemtag, "Ci", "sref", "scrJrefIprcorner", &mask, StatShr.SCRJREFIPRCORNER)
				scrJrefIprtrace = Xmlio.extractStringAttrUclc(doc, basexpath!, itemtag, "Ci", "sref", "scrJrefIprtrace", &mask, StatShr.SCRJREFIPRTRACE)
				scrJrefSrcsysinfo = Xmlio.extractStringAttrUclc(doc, basexpath!, itemtag, "Ci", "sref", "scrJrefSrcsysinfo", &mask, StatShr.SCRJREFSRCSYSINFO)
			}

			return valid
		}

		public func comm(
					_ comp: StatShr
				) -> Set<Int> {
			var items: Set<Int> = []

			if scrJrefAcqpreview == comp.scrJrefAcqpreview {items.insert(StatShr.SCRJREFACQPREVIEW)}
			if scrJrefAcqptcloud == comp.scrJrefAcqptcloud {items.insert(StatShr.SCRJREFACQPTCLOUD)}
			if scrJrefActexposure == comp.scrJrefActexposure {items.insert(StatShr.SCRJREFACTEXPOSURE)}
			if scrJrefActlaser == comp.scrJrefActlaser {items.insert(StatShr.SCRJREFACTLASER)}
			if scrJrefActservo == comp.scrJrefActservo {items.insert(StatShr.SCRJREFACTSERVO)}
			if scrJrefIprcorner == comp.scrJrefIprcorner {items.insert(StatShr.SCRJREFIPRCORNER)}
			if scrJrefIprtrace == comp.scrJrefIprtrace {items.insert(StatShr.SCRJREFIPRTRACE)}
			if scrJrefSrcsysinfo == comp.scrJrefSrcsysinfo {items.insert(StatShr.SCRJREFSRCSYSINFO)}

			return items
		}

		public func diff(
					_ comp: StatShr
				) -> Set<Int> {
			let commitems: Set<Int> = comm(comp)
			var diffitems: Set<Int> = [StatShr.SCRJREFACQPREVIEW, StatShr.SCRJREFACQPTCLOUD, StatShr.SCRJREFACTEXPOSURE, StatShr.SCRJREFACTLASER, StatShr.SCRJREFACTSERVO, StatShr.SCRJREFIPRCORNER, StatShr.SCRJREFIPRTRACE, StatShr.SCRJREFSRCSYSINFO]

			for ci in commitems {diffitems.remove(ci)}

			return diffitems
		}

	}

	/**
		* DpchEngData (full: DpchEngM2msessWzskData)
		*/
	public class DpchEngData: DpchEngWzsk {

		public static let SCRJREF: Int = 1
		public static let STATSHR: Int = 2

		public init() {
			super.init(VecWzskVDpch.DPCHENGM2MSESSWZSKDATA)
		}

		public var statshr: StatShr = StatShr()

		public override func getSrefsMask() -> String {
			var ss: Array<String> = []

			if has(DpchEngData.SCRJREF) {ss.append("scrJref")}
			if has(DpchEngData.STATSHR) {ss.append("statshr")}

			return StrMod.vectorToString(ss)
		}

		public override func readXML(
					_ doc: XMLDocument
					, _ _basexpath: String
					, _ addbasetag: Bool
				) {
			var valid: Bool
			var basexpath: String? = _basexpath

			clear()

			if addbasetag {basexpath = Xmlio.checkUclcXPaths(doc, _basexpath, "DpchEngM2msessWzskData")}
			valid = basexpath != nil

			if valid {valid = Xmlio.checkXPath(doc, basexpath!)}

			if valid {
				scrJref = Xmlio.extractStringUclc(doc, basexpath!, "scrJref", "", &mask, DpchEngData.SCRJREF)
				if statshr.readXML(doc, basexpath!, true) {add(DpchEngData.STATSHR)}
			} else {
				scrJref = ""
				statshr = StatShr()
			}
		}

	}

}
